diff -urN a/observium/includes/discovery/storage/proxmox.inc.php b/observium/includes/discovery/storage/proxmox.inc.php
--- a/observium/includes/discovery/storage/proxmox.inc.php	1970-01-01 01:00:00.000000000 +0100
+++ b/observium/includes/discovery/storage/proxmox.inc.php	2023-03-18 15:02:32.616176761 +0100
@@ -0,0 +1,39 @@
+<?php
+
+// Discover Proxmox storage.
+
+require_once("Proxmox/vendor/autoload.php");
+use Proxmox\Request;
+
+if ($config['enable_proxmox'] == '1' && $device['os'] == "pve")
+{
+  echo("Proxmox storage: ");
+
+    $configure = [
+        'hostname' => $device['hostname'],
+        'username' => $config['proxmox_username'],
+        'password' => $config['proxmox_password'] ?? null,
+        'realm' => $config['proxmox_realm'],
+        'token_name' => $config['proxmox_token_name'] ?? null,
+        'token_value' => $config['proxmox_token_value'] ?? null,
+    ];
+    Request::Login($configure);
+
+    $nodes = Request::Request('/nodes', null, 'GET');
+    foreach ($nodes->data as $node) {
+        if ($device['hostname'] == $node->node) {
+            $storages = Request::Request("/nodes/".$node->node."/storage", null, 'GET');
+            foreach ($storages->data as $val) {
+                echo $val->storage. " ";
+                discover_storage($valid['storage'], $device, $val->storage, $val->type, 'proxmox', $val->storage, 1, $val->total, $val->total - $val->avail, array());
+            }
+            echo(PHP_EOL);
+            echo(PHP_EOL);
+            return;
+        }
+    }
+
+  echo(PHP_EOL);
+  echo(PHP_EOL);
+
+}
diff -urN a/observium/includes/discovery/storage.inc.php b/observium/includes/discovery/storage.inc.php
--- a/observium/includes/discovery/storage.inc.php	2023-03-18 15:32:48.636345496 +0100
+++ b/observium/includes/discovery/storage.inc.php	2023-03-17 16:20:05.789602910 +0100
@@ -10,6 +10,9 @@
  *
  */
 
+// Always run proxmox (module checks for os suitability and enable_proxmox)
+include("includes/discovery/storage/proxmox.inc.php");
+
 // Include all discovery modules
 
 $include_dir = "includes/discovery/storage";
diff -urN a/observium/includes/discovery/virtual-machines/proxmox.inc.php b/observium/includes/discovery/virtual-machines/proxmox.inc.php
--- a/observium/includes/discovery/virtual-machines/proxmox.inc.php	1970-01-01 01:00:00.000000000 +0100
+++ b/observium/includes/discovery/virtual-machines/proxmox.inc.php	2023-03-18 15:02:17.094675653 +0100
@@ -0,0 +1,39 @@
+<?php
+
+// Try to discover Proxmox Virtual Machines.
+
+require_once("Proxmox/vendor/autoload.php");
+use Proxmox\Request;
+
+if ($config['enable_proxmox'] == '1' && $device['os'] == "pve")
+{
+  echo("Proxmox VM: ");
+
+    $configure = [
+        'hostname' => $device['hostname'],
+        'username' => $config['proxmox_username'],
+        'password' => $config['proxmox_password'] ?? null,
+        'realm' => $config['proxmox_realm'],
+        'token_name' => $config['proxmox_token_name'] ?? null,
+        'token_value' => $config['proxmox_token_value'] ?? null,
+    ];
+    Request::Login($configure);
+
+    $nodes = Request::Request('/nodes', null, 'GET');
+    foreach ($nodes->data as $node) {
+        if ($device['hostname'] == $node->node) {
+            $vms = Request::Request("/nodes/".$node->node."/qemu", null, 'GET');
+            foreach ($vms->data as $vm) {
+                echo $vm->name. " ";
+                discover_virtual_machine($valid, $device, array('id' => $vm->vmid, 'name' => $vm->name, 'cpucount' => $vm->cpus,
+                'memory' => $vm->maxmem, 'status' => $vm->status, 'type' => 'proxmox', 'source' => 'proxmox', 'protocol' => 'API'));
+            }
+        }
+    }
+
+  // Clean up removed VMs of type proxmox
+  check_valid_virtual_machines($device, $valid, 'proxmox');
+  echo(PHP_EOL);
+}
+
+// EOF
diff -urN a/observium/includes/discovery/virtual-machines.inc.php b/observium/includes/discovery/virtual-machines.inc.php
--- a/observium/includes/discovery/virtual-machines.inc.php	2023-03-17 14:37:45.955363755 +0100
+++ b/observium/includes/discovery/virtual-machines.inc.php	2023-03-17 16:07:52.508484655 +0100
@@ -14,6 +14,8 @@
 
 // Always run libvirt (module checks for os suitability and enable_libvirt)
 include("includes/discovery/virtual-machines/libvirt.inc.php");
+// Always run proxmox (module checks for os suitability and enable_proxmox)
+include("includes/discovery/virtual-machines/proxmox.inc.php");
 
 // Include all discovery modules by MIB
 $include_dir = "includes/discovery/virtual-machines";
diff -urN a/observium/includes/polling/storage/proxmox.inc.php b/observium/includes/polling/storage/proxmox.inc.php
--- a/observium/includes/polling/storage/proxmox.inc.php	1970-01-01 01:00:00.000000000 +0100
+++ b/observium/includes/polling/storage/proxmox.inc.php	2023-03-18 15:02:45.586745951 +0100
@@ -0,0 +1,36 @@
+<?php
+
+// Polling Proxmox storage.
+
+require_once("Proxmox/vendor/autoload.php");
+use Proxmox\Request;
+
+if ($config['enable_proxmox'] == '1' && $device['os'] == "pve")
+{
+    $configure = [
+        'hostname' => $device['hostname'],
+        'username' => $config['proxmox_username'],
+        'password' => $config['proxmox_password'] ?? null,
+        'realm' => $config['proxmox_realm'],
+        'token_name' => $config['proxmox_token_name'] ?? null,
+        'token_value' => $config['proxmox_token_value'] ?? null,
+    ];
+    Request::Login($configure);
+    $nodes = Request::Request('/nodes', null, 'GET');
+    foreach ($nodes->data as $node) {
+        if ($device['hostname'] == $node->node) {
+            $storages = Request::Request("/nodes/".$node->node."/storage", null, 'GET');
+            foreach ($storages->data as $val) {
+                if ($val->storage == $storage['storage_descr']) {
+                    $storage['units'] = 1;
+                    $storage['size']  = $val->total;
+                    $storage['free']  = $val->avail;
+                    $storage['used']  = $val->total - $val->avail;
+                }
+            }
+            return;
+        }
+    }
+}
+
+// EOF
